File: semah/custom_script/item/item.js
frappe.ui.form.on("Item", {
    onload: function (frm) {
        cur_frm.set_df_property('sub', 'label', 'Sub Customer');
        if (cur_frm.doc.customer_ && (!frm.is_new())) {
            frm.trigger("options")

        }
        $('[data-fieldname="bin"]').css("display", "none");
        $('[data-fieldname="update_bin"]').css("display", "none");
        $('[data-fieldname="update_all"]').css("display", "none");
        $('[data-fieldname="sub_customer"]').css("display", "none");


        cur_frm.refresh_fields();
    },
    refresh: function (frm) {
        cur_frm.set_df_property('sub', 'label', 'Sub Customer');
        $('[data-fieldname="sub_customer"]').css("display", "none");

    },
    customer_: function (frm) {
        if (cur_frm.doc.customer_) {
            frm.trigger("options")
            frappe.db.get_value('Customer', cur_frm.doc.customer_, 'customer_code')
                .then(r => {
                    var name = r.message.customer_code.toUpperCase()
                    cur_frm.set_value("item_naming", name)
                })

        }

    },
    options: function (frm) {
        frappe.db.get_doc('Customer', cur_frm.doc.customer_)
            .then(r => {
                if (r) {
                    var sub = [" ",];
                    var sub_customer = r.sub_customer
                    for (var i = 0; i < sub_customer.length; i++) {
                        sub.push(sub_customer[i].sub_customer_full_name)
                    }
                    frm.set_df_property("sub", "options", sub);
                }
            })
    },

    sub: function (frm) {
        cur_frm.set_value("sub_customer", cur_frm.doc.sub)
        if (cur_frm.doc.customer_ && cur_frm.doc.sub_customer) {
            console.log("in")
            frappe.call({
                method: "semah.custom_script.item.item.subnaming",
                async: false,
                args: {
                    customer: cur_frm.doc.customer_,
                    sub: cur_frm.doc.sub_customer
                },
                callback: function (r) {
                    if (r.message) {
                        console.log(r.message)
                        let name = r.message.toUpperCase()
                        console.log(name, r.message)
                        if (name) {
                            cur_frm.set_value("item_naming", name)
                        }
                        
                    }
                }
            })
        }
        else {
            frappe.db.get_value('Customer', cur_frm.doc.customer_, 'customer_code')
                .then(r => {
                    var name = r.message.customer_code.toUpperCase()
                    cur_frm.set_value("item_naming", name)
                })
        }

    },

    update_bin: function (frm) {
        frappe.call({
            method: "semah.custom_script.item.item.get_stock",
            async: false,
            args: {
                item: cur_frm.doc.name,
                self: cur_frm.doc
            },
            callback: function (r) {
                cur_frm.clear_table("bin_display");
                if (r) {
                    var itm = r.message
                    console.log(itm)
                    for (var i = 0; i < itm.length; i++) {
                        var rw = cur_frm.add_child("bin_display");
                        rw.warehouse = itm[i].warehouse
                        rw.batch_no = itm[i].batch_no
                        rw.bin_location = itm[i].bin_location
                        rw.expiry = itm[i].expiry
                        rw.stored_qty = itm[i].stored_qty
                        rw.stored_in = itm[i].stored_in
                        rw.length = itm[i].length
                        rw.height = itm[i].height
                        rw.width = itm[i].width
                        rw.area_use = itm[i].area_use
                        rw.sub_customer = itm[i].sub_customer
                    }

                }


                cur_frm.refresh_fields();
                frm.save()

            }
        })
    },
    update_all: function (frm) {
        frappe.call({
            method: "semah.custom_script.item.item.update_all",
            async: false,

        })
    }
})

File: semah/custom_script/item/item.py

import frappe

def validate(doc, method):
    set_item_naming(doc, method)

def on_update(self,method):
    if self.has_batch_no and self.create_new_batch:
        batch_series=self.name+"-"
        sql='''update `tabItem` set batch_number_series="{0}" where name="{1}" '''.format(batch_series,self.name)
        update= frappe.db.sql(sql,as_dict=1)
@frappe.whitelist()
def subnaming(customer,sub):
    code=""
    sql="""select item_code from `tabSub Customer` where parent="{0}" and sub_customer_full_name="{1}" """.format(customer,sub)
    subc= frappe.db.sql(sql,as_dict=1)
    customer=frappe.db.sql("""select customer_code from `tabCustomer` where name="{0}" """.format(customer),as_dict=1)
    if customer and subc:
        code=customer[0].customer_code+"-"+subc[0].item_code
    else:
        code=customer[0].customer_code
    return code




@frappe.whitelist()
def set_item_naming(doc, method):
    if doc.customer_:
        # Fetch customer code
        customer_code = frappe.db.get_value("Customer", doc.customer_, "customer_code")
        
        if customer_code:
            doc.item_naming = customer_code.upper()

        # Handle sub_customer logic
        if doc.sub_customer:
            naming = subnaming(doc.customer_, doc.sub_customer)
            if naming:
                doc.item_naming = naming.upper()




@frappe.whitelist()
def get_stock(self,item):
    stk= frappe.db.sql("""select * from `tabitem bin location` where parent="{0}"  and stored_qty !=0 """.format(item),as_dict=1)
    return stk
@frappe.whitelist()
def onload(self,item=None):
    stk= frappe.db.sql("""select DISTINCT(sub_customer),parent from `tabitem bin location` """,as_dict=1)
    sub_customer=''
    for item in stk:
        target_doc=frappe.get_doc('Item',item.parent)
        if(item.parent==self.name):
            sub_customer=sub_customer+','+item.sub_customer
            self.sub_customer=sub_customer
            self.save()

@frappe.whitelist()
def update_all(item=None):
    items= frappe.db.sql("""select name from `tabItem` """.format(item),as_dict=1)
    for k in items:
        target_doc = frappe.get_doc("Item", k.name)
        target_doc.bin_display=[]
        stk= frappe.db.sql("""select * from `tabitem bin location` where parent="{0}"  and stored_qty !=0 """.format(k.name),as_dict=1)
        for i in stk:
            rw = target_doc.append('bin_display', {})
            rw.warehouse = i.warehouse
            rw.batch_no = i.batch_no
            rw.bin_location = i.bin_location
            rw.expiry = i.expiry
            rw.stored_qty = i.stored_qty
            rw.stored_in =i.stored_in
            rw.sub_customer=i.sub_customer
            rw.area_use=i.area_use
            rw.length=i.length
            rw.height=i.height
            rw.width=i.width
            rw.manufacturing_date=i.manufacturing_date
        target_doc.save()


      
   
      